### ISSUE TITLE
I have a depth 17 tree, which I use to initialize SoftMaxTree(). Backprop gives nans and infs for gradWeight on the cuda version but the cpu version does not (seems to work fine). gradInput also blows up towards +/-inf.

### ISSUE THREAD

**Issue by pyadolla** [nicholas-leonard/cunnx#24]

**Comment by nicholas-leonard** [comment]
@pyadolla Wow that is pretty deep! 
 1. Did you try updating nnx/cunnx? 
 2. Also does your CPU code use Float or double?
 3. Do you use cloning?

**Comment by pyadolla** [comment]
@nicholas-leonard, yeah it is pretty deep I admit. (Actually it's a depth 11 tree, I originally read the depth from smt.maxDept which is computed incorrectly, I think, along with smt.maxFamilyPath [lua--nnx/SoftMaxTree.lua:116-117 should use 'nodeId' instead of 'parentId' if I'm not mistaken]).  I did update to the latest version of [lua--nnx](https://github.com/clementfarabet/lua---nnx) and cunnx before seeing the issues with exploding gradients on the gpu.

My cpu code uses float precision. As for cloning, I /also/ tried first creating smt=nn.SoftMaxTree()  with torch.setdefaulttensortype('torch.FloatTensor') on the cpu side and then smtcu=smt:clone():cuda() to get it on the gpu but I get the same problem of exploding gradients in smtcu when backprop'ing using TreeNLLCriterion()  as the loss. 

Do the tree node indices need to be ordered in a particular way when defining the 'hierarchy' table of tensors argument to nn.SoftMaxTree()?

**Comment by nicholas-leonard** [comment]
@pyadolla I just retested it with the dp packages:
```bash
th examples/recurrentlanguagemodel.lua --lstm --rho 10 --softmaxtree --cuda --dataset 'PennTreeBank' --progress --trainEpochSize 1000 --validEpochSize 1000 --trainOnly
```
I got no NaNs. The issue is not with nnx as it works for CPU.

I fixed your aforementioned SoftMaxTree like 116-117 bug though. Reinstall nnx and let me know if that fixes it.

**Comment by pyadolla** [comment]
I think I discovered what the issue is. As I mentioned before SoftMaxTree works fine on the cpu, but when using cunnx.SoftMaxTree() on the gpu I get the exploding gradients. I found that if the TreeNLLCriterion() is also placed on the gpu then the smt.gradWeight is not consistent with the cpu version. I have to clone the gradient w.r.t the input of TreeNLLCriterion() before passing it to the backward call of SoftMatTree() in order for the cpu and gpu results to match.  Below I have some code to show you what I mean, for a small depth=1 tree. I think your cunnx test doesn't pick this issue up because you give it an a priori set 'grad' vector for the backward call.

`require'cutorch'
require'nnx'
require'cunnx'

torch.setdefaulttensortype('torch.FloatTensor')
input = torch.randn(120,100)
targs = torch.IntTensor{3,4,5,6,7,8}
target = targs:index(1,torch.rand(120):mul(targs:size(1)):ceil():long())
root_id = 1
hierarchy={
    [1]=torch.IntTensor{2,7},
    [2]=torch.IntTensor{8,6,5,4,3}
}
smt = nn.SoftMaxTree(100,hierarchy,root_id)
tll = nn.TreeNLLCriterion()

inputGPU=input:cuda()
targetGPU=target:cuda()
smtGPU=smt:clone():cuda()
tllGPU=tll:clone():cuda()

smt:zeroGradParameters()
output=smt:forward{input,target}
err=tll:forward(output,target)
gradOutput=tll:backward(output,target)
smt:backward({input,target},gradOutput)
gradWeightCPU=smt.gradWeight:clone()
gradInputCPU=smt.gradInput[1]:clone()

smtGPU:zeroGradParameters()
outputGPU=smtGPU:forward{inputGPU,targetGPU}
errGPU=tllGPU:forward(outputGPU,targetGPU)
gradOutputGPU=tllGPU:backward(outputGPU,targetGPU)
smtGPU:backward({inputGPU,targetGPU},gradOutputGPU:clone())  --need to do gradOuput:clone() here for it to work right, if you remove clone() then the gpu and cpu results differ.

gradWeightGPU=smtGPU.gradWeight:clone():float()
gradInputGPU=smtGPU.gradInput[1]:clone():float()

print('CPU gradWeight min=' .. gradWeightCPU:min() ..', max=' .. gradWeightCPU:max())
print('GPU gradWeight min=' .. gradWeightGPU:min() ..', max=' .. gradWeightGPU:max())

print("GradWeight Norm Difference=" ..torch.norm(gradWeightCPU-gradWeightGPU))

print('CPU gradInput min=' .. gradInputCPU:min() ..', max=' .. gradInputCPU:max())
print('GPU gradInput min=' .. gradInputGPU:min() ..', max=' .. gradInputGPU:max())

print("GradInput Norm Difference=" ..torch.norm(gradInputCPU-gradInputGPU))
`

**Comment by nicholas-leonard** [comment]
@pyadolla That snippet really helped me find the issue. You can updated nnx to fix it. The issue was basically that gradOutput wasn't contiguous. So I added code to make it so. It took me a while to remember that clone() makes it contiguous :)

**Comment by pyadolla** [comment]
@nicholas-leonard I'm glad it helped. Funnily enough I had an inclining when I first got this error that it might have something to do with non-contiguous matrices (re-confirmed by the clone() operation). Thanks so much for finding out and fixing the problem! I'll check out the new version.
